tests/contrib/actions/email/test_email_action.py

```diff
@@ -0,0 +1,195 @@
+import pytest
+
+from spidermon.contrib.actions.email import SendEmail
+from spidermon.exceptions import NotConfigured
+
+SENDER = "sender@example.com"
+RECIPIENT = "recipient@example.com"
+SUBJECT = "Subject"
+MESSAGE = "Message"
+BODY_TEXT = "Body Text"
+BODY_HTML = "Body HTML"
+
+
+@pytest.fixture
+def email(mocker):
+    email = SendEmail(
+        sender=SENDER,
+        to=RECIPIENT,
+        fake=True,
+        subject=SUBJECT,
+        body_text=BODY_TEXT,
+        body_html=BODY_HTML,
+    )
+    email.send_message = mocker.MagicMock()
+    mocked_message = mocker.MagicMock()
+    mocked_message.as_string = mocker.MagicMock()
+    email.get_message = mocker.MagicMock(return_value=mocked_message)
+    email.render_text_template = mocker.MagicMock()
+    email.render_template = mocker.MagicMock()
+    return email
+
+
+@pytest.mark.parametrize(
+    "sender,recipient,fake,subject,exception_message",
+    [
+        (
+            None,
+            None,
+            None,
+            None,
+            "You must provide at least one recipient for the message.",
+        ),
+        (SENDER, RECIPIENT, None, None, "You must provide a subject for the message."),
+    ],
+)
+def test_init_with_invalid_params(sender, recipient, fake, subject, exception_message):
+    with pytest.raises(NotConfigured) as e:
+        SendEmail(sender=sender, to=recipient, fake=fake, subject=subject)
+    assert exception_message in str(e.value)
+
+
+def test_init_with_valid_params():
+    SendEmail(sender=None, to=None, fake=True, subject=SUBJECT)
+    SendEmail(sender=None, to=RECIPIENT, fake=False, subject=SUBJECT)
+    SendEmail(sender=SENDER, to=RECIPIENT, fake=False, subject=SUBJECT)
+
+
+def test_run_action_with_fake_present(mocker, email):
+    logger = mocker.patch("spidermon.contrib.actions.email.logger")
+    email.run_action()
+    email.get_message.assert_called_once_with()
+    logger.info.assert_called_once_with(email.get_message().as_string())
+    email.send_message.assert_not_called()
+
+
+def test_run_action_without_fake_present(mocker, email):
+    logger = mocker.patch("spidermon.contrib.actions.email.logger")
+    email.fake = False
+    email.run_action()
+    email.get_message.assert_called_once_with()
+    logger.info.assert_not_called()
+    email.send_message.assert_called_once_with(email.get_message())
+
+
+def test_get_subject_with_subject_present(email):
+    email.get_subject()
+    email.render_text_template.assert_called_once_with(SUBJECT)
```

Comment:
I think we should use the object API whenever possible instead of a mock, and leave mocks only for scenarios that cannot be tested otherwise.

For a test like this, I think it would make more sense to create an Email object, and simply check that the subject passed to the ``__init__`` method matches the one that `email.get_subject()` returns.

Reply:
My opinion is that we should isolate each part of the function to test all the cases.
For instance, in the test above, it seems that it does not matter the outcome of render_text_template (since the return of this method should be tested elsewhere), just that it is called with the correct argument.
However I'm open for discussion :)

Reply:
I agree with @Gallaecio . Isolate things is important, but in this case it seems you are mocking internal behavior of the class that can be changed anytime without affecting the purpose of the test. Usually I use mocks only when I am dealing with external objects/systems.